### PR TITLE
Threshold cache/v11

### PR DESCRIPTION
**Pull request by victorjulien**
Reimplement threshold storage and add a thread local cache, both to reduce lock contention.

https://redmine.openinfosecfoundation.org/issues/426
https://redmine.openinfosecfoundation.org/issues/6967

Replaces https://github.com/OISF/suricata/pull/11026:
- adds config opts
- doc for config opts
- address review comments

Review in isolation, no relation other thresholding PRs.

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPR1_suri_time.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPR1_stats_chk | | |
| .uptime | 645 | 672 | 104.19% | 

Pipeline 20648

File: src/util-thash.c [hunk 1]
```diff
@@ -408,6 +410,58 @@ int THashWalk(THashTableContext *ctx, THashFormatFunc FormatterFunc, THashOutput
...
+            } else {
+                THashDataUnlock(h);
+            }
+            h = next;
+        }
+        HRLOCK_UNLOCK(hb);
+    }
+
+    SCLogDebug("timeout: ending: %u entries expired", cnt);
+    return cnt;
```
Comment (inashivb) [inline src/util-thash.c:462]: Victor, I just realized we're never updating `cnt` in this fn. That looks off to me.
Apparently, it wasn't updated earlier either https://github.com/OISF/suricata/pull/11077/commits/f05b4c82ad8d76a95de72e2cc42840df88b7fd96
There are no users of `cnt` either which is why I think it has worked out ok.
Q: Do we really need to return anything from this fn?

Comment (victorjulien) [reply src/util-thash.c:462]: Fixing.

Currently the returned value isn't used indeed, but it should still be fixed.

File: src/detect-engine-threshold.c [hunk 2]
```diff
@@ -39,65 +39,413 @@
...
+
+        if (SCTIME_CMP_LT(iter->expires_at, now)) {
+            THRESHOLD_CACHE_RB_REMOVE(&threshold_cache_tree, iter);
+            HashTableRemove(threshold_cache_ht, iter, 0);
+            SCLogDebug("iter %p expired", iter);
+            cache_housekeeping_expired++;
+        }
+
+        if (cnt++ > 2)
+            break;
```
Comment (inashivb) [inline src/detect-engine-threshold.c:300]: This will run the above code 4 times now..
Is that the intention?

Comment (victorjulien) [reply src/detect-engine-threshold.c:300]: Not seeing how you'd get to 4. Can see 3 though, fixing.

**Comment by victorjulien** [comment]
Replaced by #11171

File: src/detect-engine-threshold.c (hunk 2 above)
Comment (inashivb) [reply src/detect-engine-threshold.c:300]: Like this:  http://tpcg.io/_ALOLST

Comment (victorjulien) [reply src/detect-engine-threshold.c:300]: I see, thanks.

**Pull request closed without merging**